### PR TITLE
Issues/#20 tx structure

### PR DESCRIPTION
**Pull request by r-soltani**
Adding an Equibit specific transaction structure. 
see transaction.h and transaction.cpp for major changes to the structure.
Two new fields are introduced: EQBType (1 byte) and EQBPayload (object containing an empty vector)

### CODE REVIEW

**Review by macterra — CHANGES_REQUESTED** [review]
Checks are failing

**Comment by ealefad** [comment]
there are 33 extra functional tests failing (on top of 17 disabled), should we disable or fix these tests?

**Comment by r-soltani** [comment]
Functional checks can be disabled. They will be re-enabled together with other functional tests at the Stage branch

**Review by ealefad — APPROVED** [review]

File: src/core_write.cpp [hunk 1]
```diff
@@ -163,6 +163,9 @@ void TxToUniv(const CTransaction& tx, const uint256& hashBlock, UniValue& entry,
     entry.pushKV("txid", tx.GetHash().GetHex());
     entry.pushKV("hash", tx.GetWitnessHash().GetHex());
     entry.pushKV("version", tx.nVersion);
+#ifndef BUILD_BTC
+    entry.pushKV("EQB_type", (int8_t) tx.nEQBType);
```
Comment (macterra) [inline src/core_write.cpp:167]: Why insert EQB_type here rather than just before the payload?

File: src/primitives/transaction.cpp [hunk 2]
```diff
@@ -89,9 +101,15 @@ uint256 CTransaction::GetWitnessHash() const
 }
 
 /* For backward compatibility, the hash is initialized to 0. TODO: remove the need for this default constructor entirely. */
+#ifdef BUILD_BTC
 CTransaction::CTransaction() : vin(), vout(), nVersion(CTransaction::CURRENT_VERSION), nLockTime(0), hash() {}
 CTransaction::CTransaction(const CMutableTransaction &tx) : vin(tx.vin), vout(tx.vout), nVersion(tx.nVersion), nLockTime(tx.nLockTime), hash(ComputeHash()) {}
 CTransaction::CTransaction(CMutableTransaction &&tx) : vin(std::move(tx.vin)), vout(std::move(tx.vout)), nVersion(tx.nVersion), nLockTime(tx.nLockTime), hash(ComputeHash()) {}
+#else  // BUILD_EQB
+CTransaction::CTransaction() : vin(), vout(), nVersion(CTransaction::CURRENT_VERSION), nEQBType(EQBTypes::COINBASE), nEQBPayload(), nLockTime(0), hash() {}
```
Comment (macterra) [inline src/primitives/transaction.cpp:109]: Did you consider making the COINBASE type and empty payload the defaults for the ctor?

File: src/core_write.cpp (hunk 1 above)
Comment (r-soltani) [reply src/core_write.cpp:167]: As discussed during our meeting, the type is read before parsing the vIn and vOut to be able to know the type of transactions we are expecting.

File: src/primitives/transaction.cpp (hunk 2 above)
Comment (r-soltani) [reply src/primitives/transaction.cpp:109]: yes, this method is called during coinbase creation. The transactions are defined as coinbase unless their type is explicitly defined.

File: src/core_write.cpp (hunk 1 above)
Comment (macterra) [reply src/core_write.cpp:167]: OK I'll be interested to see what difference it makes.

**Review by macterra — APPROVED** [review]

**Pull request merged**